### PR TITLE
Save the shm to the storage before restarting workers.

### LINKED ISSUE #860: The agent saves the state dict from the memory into storage if training fails.

### LINKED ISSUE #861: The agent unlinks the shared memory with signals.

### PR DESCRIPTION
**Pull request by workingloong**
### What changes were proposed in this pull request?

Save the shm to the storage before restarting workers.

### Why are the changes needed?

Fix #860 
Fix #861 

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

UT.

### CODE REVIEW

File: dlrover/python/elastic_agent/torch/training.py [hunk 1]
```diff
@@ -588,8 +590,18 @@ def _report_failure_to_master(self, failures: Dict[int, ProcessFailure]):
...
         super()._restart_workers(worker_group)
 
+    def _save_ckpt_to_storage(self):
+        """
+        The agent can save the checkpointing state dict in the shared
+        memory into the storage before restarting training processes.
+        """
+        saver: CheckpointSaver = CheckpointSaver.get_ckpt_saver()
+        if saver:
+            saver.save_shm_to_storage()
```
[the remark this answers is not in the document]
Comment (workingloong) [reply dlrover/python/elastic_agent/torch/training.py:604]: Release the lock before acquiring?

File: dlrover/python/elastic_agent/torch/ckpt_saver.py [hunk 2]
```diff
@@ -126,8 +179,10 @@ class SaverFactory(object):
 
 
 class CheckpointSaver(metaclass=ABCMeta):
```
[the remark this answers is not in the document]
Comment (workingloong) [reply dlrover/python/elastic_agent/torch/ckpt_saver.py:181]: Done.

File: dlrover/python/elastic_agent/torch/training.py [hunk 3]
```diff
@@ -499,7 +499,10 @@ def _initialize_workers(self, worker_group):
                 break
 
     def _invoke_run(self, role: str = DEFAULT_ROLE) -> RunResult:
```
[the remark this answers is not in the document]
Comment (workingloong) [reply dlrover/python/elastic_agent/torch/training.py:501]: I have added an issue #863  to do later.

**Review by liyzcj — APPROVED** [review]
lgtm

**Pull request merged**